torchacc/dist/fsdp.py

```diff
@@ -202,3 +206,179 @@ def clip_grad_norm_(self, max_grad_norm):
+            then nonzero ranks return an :class:`dict` with keys but empty value.
+        """
+        shard_meta_data = self.model.get_shard_metadata()
+        sharded_optim_state = optim.state_dict()['state']
+        optim_state_param_groups = optim.state_dict()['param_groups']
+        # unflattened and consolidated state_dict
+        consolidate_optim_state_dict: Dict[str, Any] = {
+            'state': {},
+            'param_groups': {}
+        }
+
+        # param_names(2-dim list), param_shapes(2-dim list), param_numel(2-dim list)
+        layer_name_lists, layer_size_lists, layer_numel_lists = optim_utils.get_layer_full_info(
+            shard_meta_data, self.state_dict())
+
+        # transform 2-dim list name to 1-dim list name
+        flatten_name_list = [
+            fn for layer_fn in layer_name_lists for fn in layer_fn
+        ]
+        # (rank0_only and self.model.rank == 0) or (not rank0_only)
+        if not rank0_only or self.model.rank == 0:
+            consolidate_optim_state_dict[
+                'param_groups'] = optim_state_param_groups
+            consolidate_optim_state_dict['param_groups'][0]['params'].clear()
+            for fn in flatten_name_list:
+                consolidate_optim_state_dict['param_groups'][0][
+                    'params'].append(fn)
+
+        unflat_state_dict = {fn: {} for fn in flatten_name_list}
+
+        for idx, layer_state in enumerate(sharded_optim_state.values()):
+            layer_names = layer_name_lists[idx]
+            layer_shapes = layer_size_lists[idx]
+            layer_numels = layer_numel_lists[idx]
+            for state_name, state_params in layer_state.items():
+                tensor_buffer = optim_utils.all_gather_state(
+                    state_params, self.model)
+                orig_params = optim_utils.unflatten_optim_params(
+                    tensor_buffer, layer_names, layer_shapes, layer_numels)
+
+                if not rank0_only or self.model.rank == 0:
+                    for fn, fp in zip(layer_names, orig_params):
+                        if cpu_offload:
+                            xm.mark_step()
+                            unflat_state_dict[fn][state_name] = fp.cpu()
+                        else:
+                            unflat_state_dict[fn][state_name] = fp
+                xm.mark_step()
+        consolidate_optim_state_dict['state'] = unflat_state_dict
+
+        return consolidate_optim_state_dict
+
+    def load_optim_state_dict(self,
+                              optim_state_dict: Dict[str, Any],
+                              rank0_only: bool = True) -> Dict[str, Any]:
+        """
+        Convert an optimizer state-dict so that it can be loaded into the
+        optimizer associated with the FSDP model.
+        We judge whether the optim_state_dict is sharded automatically
+                
+        Args:
+            optim_state_dict (Dict[str, Any]): The optimizer states to be loaded.
```

Comment:
Please add some comments to explain what optim_state_dict are passed by other ranks when rank0_only is True.

Reply:
There is no explanation here about how the `optim_state_dict` parameters for other ranks are passed after the optimizer state is loaded in rank 0. We can let users pass `None` instead of an empty dictionary (since users may not know what the keys are).

Reply:
I add the explanation that we will broadcast rank0's optim_state_dict info to other ranks if specified rank0_only, so it's no matter what other rank's pass in and of course other ranks can pass None.